### PR TITLE
`MeanEncoderTransform` generates wrong values

### LINKED ISSUE #490: [BUG] `MeanEncoderTransform` generates wrong values
### 🐛 Bug Report

`MeanEncoderTransform` generates wrong values

### Expected behavior

It should not.

### Code
```
from etna.datasets import TSDataset
from etna.datasets import generate_ar_df
from etna.transforms import MeanEncoderTransform
import pandas as pd
import numpy as np

df = generate_ar_df(n_segments=1, start_time="2001-01-01", periods=5)
df["target"] = [0, 1, np.NaN, 3, 4]

df_exog = generate_ar_df(n_segments=1, start_time="2001-01-01", periods=10)
df_exog.rename(columns={"target": "segment_feature"}, inplace=True)
df_exog["segment_feature"] = "segment_0"

ts = TSDataset(df=df, df_exog=df_exog, freq="D", known_future="all")

tr = MeanEncoderTransform(in_column="segment_feature", out_column="segment_mean", handle_missing="global_mean", smoothing=0)
ts.fit_transform([tr])
ts.df.loc[:, pd.IndexSlice[:, "segment_mean"]].values
```
### Output
```
array([[nan],
       [0. ],
       [0.5],
       [0. ],
       [1. ]])
```
### Expected output
```
array([[nan],
       [0. ],
       [0.5],
       [0.5],
       [1.33 ]])
```

### Environment

_No response_

### Additional context

_No response_

### Checklist

- [X] Bug appears at the latest library version

### PR DESCRIPTION
**Pull request by egoriyaa**
## Before submitting (must do checklist)

- [ ] Did you read the [contribution guide](https://github.com/etna-team/etna/blob/master/CONTRIBUTING.md)?
- [ ] Did you update the docs? We use Numpy format for all the methods and classes.
- [ ] Did you write any new necessary tests?
- [ ] Did you update the [CHANGELOG](https://github.com/etna-team/etna/blob/master/CHANGELOG.md)?

## Closing issues
closes #490

### CODE REVIEW

File: etna/transforms/encoders/mean_encoder.py [hunk 1]
```diff
@@ -214,13 +214,13 @@ def _transform(self, df: pd.DataFrame) -> pd.DataFrame:
                     # first timestamp is NaN
                     expanding_mean = y.expanding().mean().shift()
                     # cumcount not including current timestamp
-                    cumcount = y.groupby(segment_df[self.in_column].astype(str)).agg("cumcount")
+                    cumcount = segment_df.loc[y.notna()].groupby(self.in_column, dropna=False).cumcount().reindex(y.index).replace(0, np.NaN)
```
Comment (egoriyaa) [inline etna/transforms/encoders/mean_encoder.py:217]: Problem is here for `cumcount` and lower for `cumsum`

**Comment by egoriyaa** [comment]
### Per-category mode

#### segments=2000, periods=1000
2000 * 1000 unique categories
no_numba - 75 sec, numba 25

200 unique categories
no_numba - 30 sec, numba 19

1 unique category
no_numba - 18 sec, numba 18

### `MeanTransform(..., window=-1)` (for comparison)
#### segments=2000, periods=1000
7 sec

File: etna/transforms/encoders/mean_encoder.py [hunk 2]
```diff
@@ -165,6 +167,38 @@ def _count_macro_running_mean(df, n_segments):
...
+            t = target[idx]
+
+            # Mask for valid (non-NaN) target values
+            valid = ~np.isnan(t)
+
+            # Compute cumulative sums and counts for valid values
+            cumsum = np.cumsum(np.where(valid, t, 0))
+            cumcount = np.cumsum(valid).astype(np.float32)
+
+            cumsum = np.roll(cumsum, 1)
```
Comment (d-a-bunin) [inline etna/transforms/encoders/mean_encoder.py:187]: Why do we need this roll?

File: etna/transforms/encoders/mean_encoder.py [hunk 3]
```diff
@@ -254,8 +290,11 @@ def _transform(self, df: pd.DataFrame) -> pd.DataFrame:
                         .agg(["count", "sum"])
                         .reset_index()
                     )
+                    stats = stats.replace({"count": 0, "sum": 0}, np.NaN)
```
Comment (d-a-bunin) [inline etna/transforms/encoders/mean_encoder.py:293]: Why do we need this?

File: tests/test_transforms/test_encoders/test_mean_encoder_transform.py [hunk 4]
```diff
@@ -31,7 +31,7 @@ def category_ts() -> TSDataset:
 def expected_micro_category_ts() -> TSDataset:
     df = generate_ar_df(start_time="2001-01-01", periods=6, n_segments=2)
     df.rename(columns={"target": "mean_encoded_regressor"}, inplace=True)
-    df["mean_encoded_regressor"] = [np.NaN, 1, 1.5, 1.5, 2.75, 2.25] + [np.NaN, 6.0, 6.25, 7, 7.625, 8.0]
+    df["mean_encoded_regressor"] = [np.NaN, np.NaN, np.NaN, 1.5, 2.75, 2.25] + [np.NaN, np.NaN, 6.25, 7, 7.625, np.NaN]
```
Comment (d-a-bunin) [inline tests/test_transforms/test_encoders/test_mean_encoder_transform.py:34]: Why the second value is `np.NaN`? As I can see, it is a first value of category "B". 

As I understand, parameter `handle_missing` is only about missing values in `in_column`, not about first value of category or missing values in `target`.

File: tests/test_transforms/test_encoders/test_mean_encoder_transform.py [hunk 5]
```diff
@@ -114,12 +114,112 @@ def expected_ts_begin_nan_smooth_1() -> TSDataset:
...
+    df.rename(columns={"target": "regressor_mean"}, inplace=True)
+    df["regressor_mean"] = [np.NaN, np.NaN, np.NaN, np.NaN, 1.5]
+
+    ts = TSDataset(df=df, freq="D")
+
+    return ts
+
+
+@pytest.fixture
+def multiple_nan_target_old_category_ts() -> TSDataset:
```
Comment (d-a-bunin) [inline tests/test_transforms/test_encoders/test_mean_encoder_transform.py:149]: What is the difference with `multiple_nan_target_new_category_ts`? How is it old vs. new?

File: tests/test_transforms/test_encoders/test_mean_encoder_transform.py [hunk 6]
```diff
@@ -114,12 +114,112 @@ def expected_ts_begin_nan_smooth_1() -> TSDataset:
...
+    ts = TSDataset(df=df, df_exog=df_exog, freq="D", known_future="all")
+
+    return ts
+
+
+@pytest.fixture
+def expected_multiple_nan_target_old_category_ts() -> TSDataset:
+    df = generate_ar_df(n_segments=1, start_time="2001-01-01", periods=7)
+    df.rename(columns={"target": "regressor_mean"}, inplace=True)
+    df["regressor_mean"] = [np.NaN, np.NaN, np.NaN, np.NaN, 1.5, 2.75, 2.75]
```
Comment (d-a-bunin) [inline tests/test_transforms/test_encoders/test_mean_encoder_transform.py:166]: Why do we have non-NaN for first "C" value?

File: tests/test_transforms/test_encoders/test_mean_encoder_transform.py (hunk 4 above)
Comment (egoriyaa) [reply tests/test_transforms/test_encoders/test_mean_encoder_transform.py:34]: There were not timestamps with `in_column="B"` before second timestamp, so value is NaN

File: tests/test_transforms/test_encoders/test_mean_encoder_transform.py (hunk 5 above)
Comment (egoriyaa) [reply tests/test_transforms/test_encoders/test_mean_encoder_transform.py:149]: There are 2 main corner cases
1. If there is None in `target` for a new category where there were no notna `targets` yet, then the transform value should be NaN (in both cumcount and cumsum) (new)
2. If there is None in `target` for a category where there was already a notna `target`, then the transform value must be filled in with the previous one for this category (in both cumcount and cumsum) (old)

File: tests/test_transforms/test_encoders/test_mean_encoder_transform.py (hunk 6 above)
Comment (egoriyaa) [reply tests/test_transforms/test_encoders/test_mean_encoder_transform.py:166]: The last timestamp with "C" is for future, it doesn't matter what value is here

File: etna/transforms/encoders/mean_encoder.py (hunk 2 above)
Comment (egoriyaa) [reply etna/transforms/encoders/mean_encoder.py:187]: `np.cumsum` counts including current index

Comment (d-a-bunin) [reply etna/transforms/encoders/mean_encoder.py:187]: We could add comment about this.

File: tests/test_transforms/test_encoders/test_mean_encoder_transform.py (hunk 5 above)
Comment (d-a-bunin) [reply tests/test_transforms/test_encoders/test_mean_encoder_transform.py:149]: It can't be understood from name of the tests. Can't we make them more obvious or at least write a description to them?

Comment (d-a-bunin) [reply tests/test_transforms/test_encoders/test_mean_encoder_transform.py:149]: For me now it seems like `expected_multiple_nan_target_new_category_ts` is just a subset of `expected_multiple_nan_target_old_category_ts`.

File: tests/test_transforms/test_encoders/test_mean_encoder_transform.py [hunk 7]
```diff
@@ -114,12 +114,114 @@ def expected_ts_begin_nan_smooth_1() -> TSDataset:
...
-    df["mean_encoded_regressor"] = [np.NaN, np.NaN, 2 / 3, 5 / 4, 5 / 3, 2.5]
+    df["mean_encoded_regressor"] = [np.NaN, np.NaN, np.NaN, 5 / 3, 5 / 3, 2.5]
 
     ts = TSDataset(df, freq="D")
     return ts
 
 
+@pytest.fixture
+def multiple_nan_target_new_category_ts() -> TSDataset:
+    """Fixture with several timestamp with NaN target for new category where there were no notna targets yet."""
```
Comment (d-a-bunin) [inline tests/test_transforms/test_encoders/test_mean_encoder_transform.py:125]: I think it is better to write:
```
Fixture with segment having multiple NaN targets before first non-NaN target value.

Look at the segment ``A``.
```

File: tests/test_transforms/test_encoders/test_mean_encoder_transform.py [hunk 8]
```diff
@@ -114,12 +114,114 @@ def expected_ts_begin_nan_smooth_1() -> TSDataset:
...
+    df["regressor_mean"] = [np.NaN, np.NaN, np.NaN, np.NaN, 1.5]
+
+    ts = TSDataset(df=df, freq="D")
+
+    return ts
+
+
+@pytest.fixture
+def multiple_nan_target_old_category_ts() -> TSDataset:
+    """Fixture with several timestamp with NaN target for category where there was already a notna target."""
```
Comment (d-a-bunin) [inline tests/test_transforms/test_encoders/test_mean_encoder_transform.py:151]: I think it is better to write:
```
Fixture with segment having multiple NaN targets after first non-NaN target value.

Look at the segment ``B``.
```

File: tests/test_transforms/test_encoders/test_mean_encoder_transform.py (hunk 7 above)
Comment (d-a-bunin) [reply tests/test_transforms/test_encoders/test_mean_encoder_transform.py:125]: Do we really need segment `B` here?

File: tests/test_transforms/test_encoders/test_mean_encoder_transform.py (hunk 8 above)
Comment (d-a-bunin) [reply tests/test_transforms/test_encoders/test_mean_encoder_transform.py:151]: Do we really need segment `A` here?

Comment (egoriyaa) [reply tests/test_transforms/test_encoders/test_mean_encoder_transform.py:151]: My motivation was to check each case when there is more that one unique category.
Okey, let't combine these tests.

**Review by d-a-bunin — APPROVED** [review]
